core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java

```diff
@@ -638,6 +643,36 @@ public StageInfo getStageInfo(Supplier<Iterable<TaskInfo>> taskInfosSupplier)
                 failureInfo);
     }
 
+    private static List<OperatorStats> combineTaskOperatorSummaries(List<TaskInfo> taskInfos, int maxTaskOperatorSummaries)
+    {
+        // Group each unique pipelineId + operatorId combination into lists
+        Long2ObjectOpenHashMap<ArrayList<OperatorStats>> pipelineAndOperatorToStats = new Long2ObjectOpenHashMap<>(maxTaskOperatorSummaries);
+        // Expect to have one operator stats entry for each taskInfo
+        int taskInfoCount = taskInfos.size();
+        LongFunction<ArrayList<OperatorStats>> statsListCreator = key -> new ArrayList<>(taskInfoCount);
+        for (TaskInfo taskInfo : taskInfos) {
+            for (PipelineStats pipeline : taskInfo.getStats().getPipelines()) {
+                // Place the pipelineId in the high bits of the combinedKey mask
+                long pipelineKeyMask = Integer.toUnsignedLong(pipeline.getPipelineId()) << 32;
+                for (OperatorStats operator : pipeline.getOperatorSummaries()) {
+                    // Place the operatorId into the low bits of the combined key
+                    long combinedKey = pipelineKeyMask | Integer.toUnsignedLong(operator.getOperatorId());
```

Comment:
nit: `Integer.toUnsignedLong` should not be necessary. You can do `(pipeline.getOperatorId() & 0xFFFFFFFFL)` to handle negative values if necessary

Reply:
The automatic upcasting of `operator.getOperatorId()` from `int` to `long` would cause issues if the values were ever to become negative, so this is essentially the same question as the usage in `pipeline.getPipelineId()`: whether to use the constant mask as a literal or defer to the `Integer.toUnsignedLong` method implementation to make the intent (potentially) more clear. In both cases I'm happy to switch to the constant literal method if you prefer it.